OOM on S3 client, then pgwire msg-execute messages time out, and node stops responding

I'm seeing a frequent behavior in our AWS XTDB environment, after which connections don't respond and don't close.

First, at some point an OOM happens inside the S3 client:
```
2026-02-05T15:12:10.49423274Z WARN i.n.u.c.AbstractEventExecutor
A task raised an exception. Task: software.amazon.awssdk.http.nio.netty.internal.nrs.HandlerSubscriber$1@7eea4872
java.lang.OutOfMemoryError: Java heap space
	at software.amazon.awssdk.utils.BinaryUtils.copyBytesFrom(BinaryUtils.java:302)
	at software.amazon.awssdk.services.s3.internal.checksums.ChecksumCalculatingAsyncRequestBody$ChecksumCalculatingSubscriber.onNext(ChecksumCalculatingAsyncRequestBody.java:84)
	at software.amazon.awssdk.services.s3.internal.checksums.ChecksumCalculatingAsyncRequestBody$ChecksumCalculatingSubscriber.onNext(ChecksumCalculatingAsyncRequestBody.java:60)
	at software.amazon.awssdk.core.internal.async.ByteBuffersAsyncRequestBody$1.request(ByteBuffersAsyncRequestBody.java:98)
	at software.amazon.awssdk.http.nio.netty.internal.nrs.HandlerSubscriber.maybeRequestMore(HandlerSubscriber.java:303)
	at software.amazon.awssdk.http.nio.netty.internal.nrs.HandlerSubscriber.maybeStart(HandlerSubscriber.java:238)
	at software.amazon.awssdk.http.nio.netty.internal.nrs.HandlerSubscriber.provideSubscription(HandlerSubscriber.java:225)
	at software.amazon.awssdk.http.nio.netty.internal.nrs.HandlerSubscriber.access$000(HandlerSubscriber.java:41)
	at software.amazon.awssdk.http.nio.netty.internal.nrs.HandlerSubscriber$1.run(HandlerSubscriber.java:213)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:148)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:141)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:507)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:566)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:1073)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at java.base/java.lang.Thread.runWith(Thread.java:1596)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

Some minutes after, queries start to timeout. I have added some more data to the logs - we can see a `msg-execute` was being handled.
```
2026-02-05T15:20:36.093323661Z ERROR xtdb.pgwire
Uncaught exception processing message : {:cid 92, :msg {:stmt-name "", :query "SHOW LATEST_SUBMITTED_TX", :param-oids [], :msg-name :msg-parse}}
xtdb.error.Fault: Timed out waiting for 60000 ms
	at xtdb.error$fault.invokeStatic(error.clj:70)
	at xtdb.error$fault.invoke(error.clj:62)
	at xtdb.error$eval1034$fn__1035.invoke(error.clj:184)
	at xtdb.error$eval961$fn__962$G__952__969.invoke(error.clj:14)
	at xtdb.pgwire$send_ex.invokeStatic(pgwire.clj:277)
	at xtdb.pgwire$send_ex.doInvoke(pgwire.clj:271)
	at clojure.lang.RestFn.invoke(RestFn.java:445)
	at xtdb.pgwire$handle_msg.invokeStatic(pgwire.clj:1736)
	at xtdb.pgwire$handle_msg.invoke(pgwire.clj:1712)
	at xtdb.pgwire$conn_loop.invokeStatic(pgwire.clj:1767)
	at xtdb.pgwire$conn_loop.invoke(pgwire.clj:1738)
	at xtdb.pgwire$connect.invokeStatic(pgwire.clj:1814)
	at xtdb.pgwire$connect.invoke(pgwire.clj:1771)
	at xtdb.pgwire$accept_loop$fn__29461$fn__29462$fn__29463.invoke(pgwire.clj:1860)
	at clojure.lang.AFn.run(AFn.java:22)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: kotlinx.coroutines.TimeoutCancellationException: Timed out waiting for 60000 ms
	at kotlinx.coroutines.TimeoutKt.TimeoutCancellationException(Timeout.kt:189)
	at kotlinx.coroutines.TimeoutCoroutine.run(Timeout.kt:157)
	at kotlinx.coroutines.EventLoopImplBase$DelayedRunnableTask.run(EventLoop.common.kt:505)
	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:263)
	at kotlinx.coroutines.DefaultExecutor.run(DefaultExecutor.kt:105)
	... 1 common frames omitted
```

Then more `msg-execute` for the same connection keep timing out (now after 30s instead of 60s). The following is seen every 90s, 15 times, for the same connection 92 (cid=92):
```
2026-02-05T15:22:06.109241847Z
Uncaught exception processing message : {:cid 92, :msg {:portal-name "", :limit 0, :msg-name :msg-execute}}
xtdb.error.Fault: Timed out waiting for 30000 ms
	at xtdb.error$fault.invokeStatic(error.clj:70)
	at xtdb.error$fault.invoke(error.clj:62)
	at xtdb.error$eval1034$fn__1035.invoke(error.clj:184)
	at xtdb.error$eval961$fn__962$G__952__969.invoke(error.clj:14)
	at xtdb.pgwire$send_ex.invokeStatic(pgwire.clj:277)
	at xtdb.pgwire$send_ex.doInvoke(pgwire.clj:271)
	at clojure.lang.RestFn.invoke(RestFn.java:445)
	at xtdb.pgwire$handle_msg.invokeStatic(pgwire.clj:1736)
	at xtdb.pgwire$handle_msg.invoke(pgwire.clj:1712)
	at xtdb.pgwire$conn_loop.invokeStatic(pgwire.clj:1767)
	at xtdb.pgwire$conn_loop.invoke(pgwire.clj:1738)
	at xtdb.pgwire$connect.invokeStatic(pgwire.clj:1814)
	at xtdb.pgwire$connect.invoke(pgwire.clj:1771)
	at xtdb.pgwire$accept_loop$fn__29461$fn__29462$fn__29463.invoke(pgwire.clj:1860)
	at clojure.lang.AFn.run(AFn.java:22)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: kotlinx.coroutines.TimeoutCancellationException: Timed out waiting for 30000 ms
	at kotlinx.coroutines.TimeoutKt.TimeoutCancellationException(Timeout.kt:189)
	at kotlinx.coroutines.TimeoutCoroutine.run(Timeout.kt:157)
	at kotlinx.coroutines.EventLoopImplBase$DelayedRunnableTask.run(EventLoop.common.kt:505)
	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:263)
	at kotlinx.coroutines.DefaultExecutor.run(DefaultExecutor.kt:105)
	... 1 common frames omitted
```

--- a/core/src/main/kotlin/xtdb/api/storage/ObjectStore.kt
+++ b/core/src/main/kotlin/xtdb/api/storage/ObjectStore.kt
@@ -68,6 +68,8 @@ interface ObjectStore : AutoCloseable {
 
     /**
      * Stores an object in the object store.
+     *
+     * The provided ByteBuffer must not be modified during the execution of this method.
      */
     fun putObject(k: Path, buf: ByteBuffer): CompletableFuture<Unit>
 

--- a/gradle/libs.versions.toml
+++ b/gradle/libs.versions.toml
@@ -2,7 +2,7 @@
 antlr = "4.13.2"
 arrow = "18.3.0"
 arrow-adbc = "0.21.0"
-aws-sdk = "2.25.50"
+aws-sdk = "2.41.27"
 azure = "1.13.1"
 azure-storage = "12.27.0"
 clojure = "1.12.0"

--- a/modules/aws/src/main/kotlin/xtdb/aws/S3.kt
+++ b/modules/aws/src/main/kotlin/xtdb/aws/S3.kt
@@ -100,7 +100,7 @@ class S3(
                     val contentLength = buf.remaining().toLong()
                     val partNum = idx + 1
 
-                    val partResp = client.uploadPart(AsyncRequestBody.fromByteBuffer(buf)) {
+                    val partResp = client.uploadPart(AsyncRequestBody.fromByteBufferUnsafe(buf)) {
                         it.bucket(bucket)
                         it.key(s3Key)
                         it.uploadId(uploadId)
@@ -179,7 +179,7 @@ class S3(
 
             val contentLength = buf.remaining().toLong()
 
-            client.putObject(AsyncRequestBody.fromByteBuffer(buf)) {
+            client.putObject(AsyncRequestBody.fromByteBufferUnsafe(buf)) {
                 it.bucket(bucket)
                 it.key(s3Key)
                 it.contentLength(contentLength)

